### ISSUE TITLE
add a slatedb admin tool

### ISSUE THREAD

**Issue by rodesai** [slatedb/slatedb#151]
**Is your feature request related to a problem? Please describe.**
Currently there is no way to execute admin actions against a db instance. For example, an administrator may want to read the manifest, read db rows, take a snapshot, manually execute a compaction, etc. Currently the only option is to write your own client code to do this.

**Describe the solution you'd like**
An admin CLI tool that lets you take admin actions, e.g.

```bash
$ slatedb --provider aws --bucket foo --region us-west-2 --path db read <read args>
$ slatedb --provider aws --bucket foo --region us-west-2 --path db read-manifest <read manifest args>
$ slatedb --provider aws --bucket foo --region us-west-2 --path db snapshot <snapshot args>
$ slatedb --provider aws --bucket foo --region us-west-2 --path db compact <compact args>
```

**Describe alternatives you've considered**

**Additional context**
This was originally suggested by @criccomini in the feedback to this PR which adds compaction against a real db to the compaction bench tool: https://github.com/slatedb/slatedb/pull/149#pullrequestreview-2259349138. That feature makes more sense in the context of an admin tool.

clap looks like a great lib for doing CLI parsing: https://docs.rs/clap/latest/clap/
I used it for the db-bench tool here: #148

**Comment by criccomini** [comment]
+1 for CLAP. Seems to be the go-to for CLI in Rust.

**Comment by Hrushi20** [comment]
Hey! I wanna work on this. Is this an urgent requirement?

**Comment by criccomini** [comment]
@Hrushi20 I wouldn't say it's urgent. But it's definitely nice to have, and is something we absolutely will need in the future.

**Comment by criccomini** [comment]
@Hrushi20 I've assigned to you. If you'd rather work on something else, just LMK. :)

**Comment by Hrushi20** [comment]
Hey! Can I get a brief overview of what each command should do? Also do we need to have an authentication mechanism for admins to use cli?

**Comment by criccomini** [comment]
> Can I get a brief overview of what each command should do

Let's start *very* simple. We can add other commands in follow-on. Let's do _only_ `slatedb compact`. You can move the `load_compaction_as_job` from the compaction benchmark tool into the admin tool.

Once you you have the CLI implemented with the `compact` command, we can add other commands like Rohan lists. :)

> do we need to have an authentication mechanism for admins to use cli?

Nope! The auth is already handled by S3. If the user has read access, they can read. If they have write access, they can write.

**Comment by Hrushi20** [comment]
I've created a basic Cli tool. Do I use `TokioCompactionExecutor` to execute the compaction process?

**Comment by criccomini** [comment]
@Hrushi20 I think you should be able to copy the logic from the `compaction_execute_bench.rs` file. Rohan is using `TokioCompactionExecutor` in there (inside `run_bench`). I think using the Tokio executor is fine for the CLI as well.

**Comment by agavra** [comment]
@Hrushi20 hello! I wanted to check up on your progress on this issue, I didn't realize there was existing work on it and created #207 - I'm happy to incorporate any work you've already done, or review your work if you've got a draft PR. Keep me posted on your progress!

**Comment by criccomini** [comment]
Spoke with @Hrushi20 on discord. He's on vacation right now and said he hasn't made much progress. Moving this issue to @agavra  since he's expressed interest in working on it.

**Comment by agavra** [comment]
### Some Thoughts

So turns out an admin tool is a bit more complicated than it seems at first glance. Here's a summary of some thoughts I had:

1. (**fencing**) It's important that the admin tool does not disrupt active applications. Since SlateDB has a fencing mechanism, it's possible that embedding an instance of SlateDB in the admin tool would cause active applications to be fenced. In some scenarios (i.e. with a stream processor) that could cause downstream effects that are non-negligible in impact, such as shuffling task allocation and re-bootstrapping caches.
2. (**use cases**) Because of (1), we can split the admin tool into two use cases: introspection & modification. Introspection are read-only operations such as reading the manifest file and gathering statistics. Modification is further split into two categories: (2A) nonfunctional tasks (e.g. compaction) and (2B) functional changes (e.g. inserting data).
3. (**introspection**) Reads should never fence other active instances and can connect to the same object store buckets and read committed data. There are some introspection operations that are lightweight (e.g. reading the manifest) and can be done without instantiating an entire instance of SlateDB and then there are use cases that are heavyweight (e.g. reading a row) that would require bootstrapping the WAL, indexes etc... For the latter use cases, it probably makes sense to have some sort of long-running background task whether that's designing it as a REPL (c.f. #207) or allowing it to run as a daemon.
4. (**modification**) This is where it gets interesting. I think the compaction (2A) was designed to be fenced separately from the main client, but as part of this work we should definitely make sure that's the case (i.e. if compaction runs somewhere other than the main process _and_ fences the compaction on the main process, the put/get/delete operations still work). Writes (2B) are by design only single-writer, so I think it makes sense for the admin client to not support writes (we don't want to get into a situation where there's a fencing war between the active application and the admin instance).
5. (**object store backdoor**) When executing long-lived compaction jobs, we could design it in two ways: (a) we can run an embedded instance of SlateDB that runs compaction with the main client disabled, (b) we can use an object-store backdoor to communicate to an already running process of SlateDB. There are pros and cons of each (see below).
6. (**reviving in-band compaction**) If an admin client fences the "main" long-lived compaction process, there should be some mechanism to signal to the original compaction process that it is good to resume once the admin completes - otherwise we risk fencing the admin tool if it is configured to automatically restart in situations it gets fenced.

------

### OS Backdoor Approach (Rejected)

I wanted to record this thought somewhere so we don't lose it, but I think we should shelf it for now:
 
If we decide to run an embedded instance of SlateDB for the admin tool, the process that kicks off the job, such as an oncall engineer's laptop, will need to stay awake during the entire duration -- which for a significant lower-level compaction job can take significant amount of time. The machine that kicks off the job also needs to be fully privileged to access the object store. 

With a backdoor approach, we could create an object in the object store that the main instance of SlateDB polls (unfortunately AFAIK we can't rely on the presence of some triggering mechanism as each object store integrates in different ways) at some interval and execute that command.

The three main pros of a backdoor approach:
1. we still have only a single instance that's running any modifications. This means there's no risk of fencing
2. the admin instance does not need to be awake/blocking for the entire time an operation is happening
3. the admin instance does not need full security privileges and can only require a limited, easily auditable set

There are downsides, however, so I recommend we reject this alternative for now unless the requirements for the above become necessary:
1. There's no standard triggering mechanism for events across object stores, so polling is required
2. It introduces complexity in the main client (though the blast radius is limited to the compaction loop)

Lastly, it is simple to build in an event-driven manner by users of SlateDB once we expose an API to compact on SlateDB.

**Comment by criccomini** [comment]
Thanks for writing this up. It makes a lot of sense.

I don't think we should support puts/deletes. I feel apps should build that in at their layer if they want to modify the DB. They can route gets/puts through their writers or live with fencing the writer. Regarding the OS backdoor approach, I had the same thought and the same concerns. I'm inclined to table it for now. Perhaps if compaction proves tricky or we find a need for modification, we can revisit it.

As for read functionality, I can imagine wanting to:

1. List all manifests
2. Dump a manifest
3. Dump an SST's stats (insertion time, bloom filters, key count, size, index/block info, writer_epoch, and so on)
4. List orphaned SSTs (not in any manifest and not in WAL) and their last modify time 
5. List keys (with optional range)
6. Get current value for a key
7. Get all values for a key (along with which SST each value is in)

Regarding a daemon or REPL, I think such a thing would be useful for (5), (6), and (7). In the interest of doing things simply, incrementally, and empirically, I would prefer to start with the simplest thing: no daemon/REPL. The CLI can start the DB each time it runs. If/when we get fed up with slow start times due to long WALs/large indexes, we can add a daemon/REPL.

For modification, I'd like to be able to:

1. Trigger garbage collection
2. Trigger compaction

Of the above list, I think only compaction requires some thought around fencing. We have `compactor_epoch` in the manifest. If we trigger a compaction, I think we need to bump that epoch. In doing so, I think we'd fence the main client's compactor, which I _think_ will cause the compactor to die permanently. Once #111 is done, it'll probably fail the whole client. We'll need to think about how we want to deal with that.

I think this set of operations would be quite useful. I'll think a bit about compaction as I think that's the really tricky one.

tl;dr My vote is to start with the read functionality (introspection) and skip daemon/REPL for now.

**Comment by rodesai** [comment]
I mostly agree @criccomini. Couple additional thoughts:

> I don't think we should support puts/deletes

I think there's use-cases where there isn't a long-lived active writer for the cli to fence, and it'd be nice to support putting data into the db. The most obvious one is someone just looking for something to run when getting started with slatedb - it'd be nice to just have a cli they can point at a bucket and do some puts/gets against. That's a rather different use-case from an admin tool, but if the tool is already able to read keys, then it's pretty close to what it's already doing and it seems straightforward to include it by requiring the user to specify some flag like `--mode writer`. Alternatively, we can have the admin tool be strictly operational stuff (which I think includes 1-4 and 7 from your introspection list and your modification list), and put stuff that reads/writes data into a separate tool. I'm leaning towards the former right now since it feels simpler to just have everything in one place.

>  For modification, I'd like to be able to:
I'd also add:
- Initialize an empty db
- Create a snapshot of an existing db

For the compactor, I agree it's tricky. If there's a compactor running, and you want to manually schedule a compaction, then it should really be the thing that accepts that compaction and drives it because the compactor needs to know about running compactions so it can avoid scheduling conflicting ones. So I don't see a great way around adding some kind of backdoor - we need some way to signal to the running compactor that it should run the compaction we want, or alternatively that the compaction we want is running in our tool and what it's current status is.

**Comment by criccomini** [comment]
> I'm leaning towards the former right now since it feels simpler to just have everything in one place.

+1 I agree. We don't have to call the CLI the admin tool. It can just be the CLI, which can be used for admin operations, newb/experimentation operations, debugging, etc.

> So I don't see a great way around adding some kind of backdoor - we need some way to signal to the running compactor that it should run the compaction we want, or alternatively that the compaction we want is running in our tool and what it's current status is.

I'm wondering if this is a good time for us to add compactor plan persistence to SlateDB? We had talked about doing this eventually--making the compactor store what it's doing so it can resume on failure. If we had such a feature, might we be able to leverage it to trigger a future compaction? Rather than having the CLI run a compaction, it could write a job plan to the same spot the compactor does (in the same format), and the compactor will pick it up when it runs (treating it like a recovery). Something like that...

**Comment by criccomini** [comment]
Closing this since we now have an admin tool. =)
